### ISSUE TITLE
Crossgen hangs and crashes on MacOs

### ISSUE THREAD

**Issue by Tratcher** [dotnet/runtime#81570]
### Description

dotnet/AspNetCore's main branch has been experiencing hangs in crossgen on mac since at least 12/26/22. While we haven't been able to capture a dump from those, we did recently get a dump from a segfault in the same area.

Hangs:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=156458&view=logs&j=9895f7d8-b062-540b-461c-07f8695bfcc6&t=a6c33838-4cf4-594d-88cb-801ae7f8db2f&l=533
https://dev.azure.com/dnceng-public/public/_build/results?buildId=145996&view=logs&j=f43b187f-6c15-531c-a59b-e4fba4e9ed20&t=98a09275-a4dd-593e-c21a-1f8bea3be5b1&l=2471
https://dev.azure.com/dnceng-public/public/_build/results?buildId=141806&view=logs&j=f43b187f-6c15-531c-a59b-e4fba4e9ed20&t=98a09275-a4dd-593e-c21a-1f8bea3be5b1&l=2371
https://dev.azure.com/dnceng-public/public/_build/results?buildId=136865&view=logs&j=8e9e2812-09db-5a5d-c249-d96adf9f47fa&t=51e8c4b1-e760-5f88-6965-6fecd74b90ed&l=535
https://dev.azure.com/dnceng-public/public/_build/results?buildId=135884&view=logs&j=f43b187f-6c15-531c-a59b-e4fba4e9ed20&t=98a09275-a4dd-593e-c21a-1f8bea3be5b1&l=2483
https://dev.azure.com/dnceng-public/public/_build/results?buildId=130679&view=logs&j=8e9e2812-09db-5a5d-c249-d96adf9f47fa&t=51e8c4b1-e760-5f88-6965-6fecd74b90ed&l=543
https://dev.azure.com/dnceng-public/public/_build/results?buildId=129390&view=logs&j=8e9e2812-09db-5a5d-c249-d96adf9f47fa&t=51e8c4b1-e760-5f88-6965-6fecd74b90ed&l=533
https://dev.azure.com/dnceng-public/public/_build/results?buildId=127806&view=logs&j=9895f7d8-b062-540b-461c-07f8695bfcc6&t=a6c33838-4cf4-594d-88cb-801ae7f8db2f&l=531
https://dev.azure.com/dnceng-public/public/_build/results?buildId=125508&view=logs&j=9895f7d8-b062-540b-461c-07f8695bfcc6&t=a6c33838-4cf4-594d-88cb-801ae7f8db2f&l=565

Crash:
https://dev.azure.com/dnceng-public/public/_build/results?buildId=157320&view=results
https://dev.azure.com/dnceng-public/_apis/resources/Containers/35147423/MacOS_Test_Dumps?itemPath=MacOS_Test_Dumps%2Fdotnet-50140.1675351152.core

We'll collect additional details in this issue as we find them.

cc: @BrennanConroy @dougbu 

### Reproduction Steps

dotnet/AspNetCore CI for the main branch on mac.
https://dev.azure.com/dnceng-public/public/_build?definitionId=83&_a=summary

### Expected behavior

No hangs or crashes

### Actual behavior

Hangs and/or crashes

### Regression?

Yes

### Known Workarounds

Re-run the job

### Configuration

.NET 8

**Comment by mangod9** [comment]
@trylek @janvorli FYI.

**Comment by janvorli** [comment]
The error message of the System.InvalidOperationException in the dump case says:
`Cannot read keys when either application does not have a console or when console input has been redirected. Try Console.Read.`

The managed call stack is
```
00007FFEE40C0F80 0000000115B58CD4 System.ConsolePal.ReadKey(Boolean)
00007FFEE40C0FC0 00000001159C0AC2 IStartupInjectionAssemblyName.Program.Main(System.String[])
```

I'm not sure why the exception message / stack trace was not written to the stderr, maybe it is the same underlying issue that caused the actual exception.

**Comment by BrennanConroy** [comment]
Why is crossgen even touching that project? It's a test project located at https://github.com/dotnet/aspnetcore/blob/main/src/Hosting/test/testassets/IStartupInjectionAssemblyName/Program.cs and the crossgen args don't look like they have anything to do with that project:
> "/Users/runner/work/1/s/.packages/microsoft.netcore.app.crossgen2.osx-x64/8.0.0-preview.2.23101.9/tools/crossgen2" --targetarch:x64 --targetos:osx -O @"/Users/runner/work/1/s/artifacts/obj/Microsoft.AspNetCore.App.Runtime/Release/net8.0/osx-x64/crossgen/PlatformAssembliesPathsCrossgen2.rsp" "-m:/Users/runner/work/1/s/.packages/microsoft.netcore.app.runtime.osx-x64/8.0.0-preview.2.23101.9/tools/StandardOptimizationData.mibc" -o:"/Users/runner/work/1/s/artifacts/bin/Microsoft.AspNetCore.App.Runtime/Release/net8.0/osx-x64/System.Security.Cryptography.Xml.dll" "/Users/runner/work/1/s/.packages/system.security.cryptography.xml/8.0.0-preview.2.23101.9/lib/net8.0/System.Security.Cryptography.Xml.dll"

**Comment by janvorli** [comment]
The dump file is not from crossgen2 as you can see from the stack trace. The application start method is `IStartupInjectionAssemblyName.Program.Main`

**Comment by trylek** [comment]
For the role of Crossgen2 @ivdiazsa is probably the best to comment as he was one of the two people implementing Crossgen2 in ASP.NET, the other person being @davidwrighton who's currently on parental leave.

**Comment by mangod9** [comment]
@Tratcher does this still need investigation? From what Jan found it appears this might not be related to cg2?

**Comment by Tratcher** [comment]
The crash apparently wasn't, but we're still suffering from the regular hangs and could use some help to investigate those. Unfortunately we haven't been able to capture any dumps from the hangs.

**Comment by mangod9** [comment]
does it only repro on CI machines or can we setup a local repro for it?

**Comment by Tratcher** [comment]
We've only managed to repro it on the CI.

**Comment by captainsafia** [comment]
@mangod9 @amcasey and I are taking a look at this. Not sure if it is a helpful detail, but the issue appears to only repro on x64 macOS. The arm64 macOS build does not hang on the crossgen build.

**Comment by mangod9** [comment]
Ok interesting. Thanks for additional info -- so is it reproing locally on x64, or still in CI?

**Comment by BrennanConroy** [comment]
> The arm64 macOS build does not hang on the crossgen build.

https://dev.azure.com/dnceng-public/public/_build/results?buildId=147425&view=logs&j=8e9e2812-09db-5a5d-c249-d96adf9f47fa&t=51e8c4b1-e760-5f88-6965-6fecd74b90ed

**Comment by amcasey** [comment]
@mangod9 Just an (apparently incorrect) observation from historical CI runs.

**Comment by BrennanConroy** [comment]
Got the issue to repro locally and got a dump. Interestingly enough, just attaching and unattaching lldb fixes the hang.

Someone ping me offline so I can get the 2GB dump file to you.

**Comment by janvorli** [comment]
From the dump shared by @BrennanConroy, I can see that the hang occurred even before the runtime initialization completed. The GC initialization is starting a GC thread and it waits on an event signaling that the thread started. That event is set at the very beginning of the thread function. The process has four threads at that point, one is starting a thread and three are ones started using the same mean. Since we don't have symbols for the crossgen2 (it is a nativeaot build of crossgen2), I don't see any locals and I cannot tell whether the problem is that a thread has not started at all or if the event signaling failed.
```
(lldb) bt all
* thread #1
  * frame #0: 0x00007fff703ff882 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff704c0425 libsystem_pthread.dylib`_pthread_cond_wait + 698
    frame #2: 0x000000010f78ae7b crossgen2`UnixEvent::Wait(unsigned int) + 155
    frame #3: 0x000000010f7379f9 crossgen2`CLREventStatic::Wait(unsigned int, bool, bool) + 121
    frame #4: 0x000000010f73b8e1 crossgen2`GCToEEInterface::CreateThread(void (*)(void*), void*, bool, char const*) + 81
    frame #5: 0x000000010f7ab987 crossgen2`SVR::gc_heap::init_gc_heap(int) + 3607
    frame #6: 0x000000010f7aab46 crossgen2`SVR::gc_heap::make_gc_heap(SVR::GCHeap*, int) + 438
    frame #7: 0x000000010f7dcb26 crossgen2`SVR::GCHeap::Initialize() + 2902
    frame #8: 0x000000010f73ae3b crossgen2`RedhawkGCInterface::InitializeSubsystems() + 251
    frame #9: 0x000000010f73fa71 crossgen2`RhInitialize + 305
    frame #10: 0x000000010f7359d5 crossgen2`main + 21
    frame #11: 0x00007fff702bbcc9 libdyld.dylib`start + 1
  thread #2
    frame #0: 0x00007fff703ff882 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff704c0425 libsystem_pthread.dylib`_pthread_cond_wait + 698
    frame #2: 0x000000010f78c1fb crossgen2`GCEvent::Impl::Wait(unsigned int, bool) + 155
    frame #3: 0x000000010f7a293f crossgen2`SVR::gc_heap::gc_thread_function() + 159
    frame #4: 0x000000010f7a2896 crossgen2`SVR::gc_heap::gc_thread_stub(void*) + 54
    frame #5: 0x000000010f73bafe crossgen2`GCToEEInterface::CreateThread(void (*)(void*), void*, bool, char const*)::$_0::__invoke(void*) + 78
    frame #6: 0x00007fff704c0109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff704bbb8b libsystem_pthread.dylib`thread_start + 15
  thread #3
    frame #0: 0x00007fff703ff882 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff704c0425 libsystem_pthread.dylib`_pthread_cond_wait + 698
    frame #2: 0x000000010f78c1fb crossgen2`GCEvent::Impl::Wait(unsigned int, bool) + 155
    frame #3: 0x000000010f7a2911 crossgen2`SVR::gc_heap::gc_thread_function() + 113
    frame #4: 0x000000010f7a2896 crossgen2`SVR::gc_heap::gc_thread_stub(void*) + 54
    frame #5: 0x000000010f73bafe crossgen2`GCToEEInterface::CreateThread(void (*)(void*), void*, bool, char const*)::$_0::__invoke(void*) + 78
    frame #6: 0x00007fff704c0109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff704bbb8b libsystem_pthread.dylib`thread_start + 15
  thread #4
    frame #0: 0x00007fff703ff882 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fff704c0425 libsystem_pthread.dylib`_pthread_cond_wait + 698
    frame #2: 0x000000010f78c1fb crossgen2`GCEvent::Impl::Wait(unsigned int, bool) + 155
    frame #3: 0x000000010f7a2911 crossgen2`SVR::gc_heap::gc_thread_function() + 113
    frame #4: 0x000000010f7a2896 crossgen2`SVR::gc_heap::gc_thread_stub(void*) + 54
    frame #5: 0x000000010f73bafe crossgen2`GCToEEInterface::CreateThread(void (*)(void*), void*, bool, char const*)::$_0::__invoke(void*) + 78
    frame #6: 0x00007fff704c0109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff704bbb8b libsystem_pthread.dylib`thread_start + 15
```

Since the issue occurs so early, I have tried to just run it in a loop without any arguments on my local x64 mac, but I couldn't repro the issue even in thousands of iterations.  It is possible that the issue is related to the macOS version somehow. My device has macOS 12 while the CI one has macOS 11.

I have even tried to use the same binary of crossgen2 the dump is from and I still couldn't repro it. I have also tried to run it on my mac M1 under the Rosetta 2 emulation and there was no repro either.

I'll try a couple of things, but I might need to grab a machine from the CI to try to repro it there.

**Comment by MichalStrehovsky** [comment]
> Since we don't have symbols for the crossgen2 (it is a nativeaot build of crossgen2)

You weren't seeing locals for the native code? That's not expected - I see in the backtraces above that we do have _some_ debug information (the method signatures show parameter types) - we should have the rest as well. Could you please file an issue with what you saw? Could it be just the usual trouble with optimized code?

**Comment by janvorli** [comment]
@MichalStrehovsky there were symbols for methods, but not for locals anywhere.

**Comment by janvorli** [comment]
Let me check what symbol info the crossgen2 had. When I build it locally, it contained all the debug symbols including locals. It  is possible that the official build had just regular symbols, but not debug ones. That would explain what I was seeing.

**Comment by janvorli** [comment]
Hmm, the size of the file is about the same as the one I've built locally, so I guess the symbols are likely there  and  it was a quirk of lldb  w.r.t. a core file.

**Comment by MichalStrehovsky** [comment]
> Hmm, the size of the file is about the same as the one I've built locally, so I guess the symbols are likely there and it was a quirk of lldb w.r.t. a core file.

Thanks for checking (and for root causing and fixing the original issue)! Btw, should we close this issue now that #82709 merged?

**Comment by janvorli** [comment]
I have left it open since the title says "Crossgen hangs **and crashes** on macOS". The hang is fixed by #82709, but it is unrelated to any crash.
I am going to look at the dump of the crash that is fortunately still there and see where it stems from.

**Comment by janvorli** [comment]
Ah, the crash is what I have already looked into before and commented above: https://github.com/dotnet/runtime/issues/81570#issuecomment-1416183263

I am closing this issue as that one has nothing to do with crossgen2 and seems to be a problem with a test.
